apps/settings/js/modules/bluetooth/bluetooth_connection_manager.js

```diff
@@ -0,0 +1,408 @@
+     * The method will disconnect current connected device first.
+     * Then, it will start connecting a paired device with the device's adapter.
+     *
+     * @access public
+     * @memberOf BluetoothConnectionManager
+     * @param {Object BluetoothDevice} device
+     * @return {Promise}
+     */
+    connect: function btcm_connect(device) {
+      if (!this._defaultAdapter) {
+        return Promise.reject('default adapter is not existed!!');
+      }
+
+      // In code base v1, it always stopDiscovery before connect with a device.
+      // Shall we stopDiscovery here too?
+
+      // Disconnect current connected device first.
+      Debug('connect(): Want to connect device address = ' + device.address +
+            ', name = ' + device.name);
+      return this._disconnectCurrentConnectedDevice().then(() => {
+        // Have checked there is no connected device.
+        // Or have disconnected current connected device.
+        // We can try to connect with other device.
+        return this._connect(device).then(() => {
+          Debug('connect(): Resolved');
+        }, (reason) => {
+          Debug('connect(): reason = ' + reason);
+          return Promise.reject(reason);
+        });
+      });
+    },
+
+    /**
+     * The method will connect the input device with the device's adapter.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @param {Object BluetoothDevice} device
+     * @returns {Promise}
+     */
+    _connect: function btcm__connect(device) {
+      return new Promise((resolve, reject) => {
+        if (!this._defaultAdapter) {
+          reject('default adapter is not existed!!');
+        }
+
+        // Save the connecting address
+        this.connectingAddress = device.address;
+        Debug('_connect(): start connecting device = ' + device);
+        var req = this._defaultAdapter.connect(device);
+        req.onsuccess = () => {
+          // At least one profile is connected.
+          // Clean up the connecting address
+          this.connectingAddress = null;
+          // Save the connected address
+          this.connectedAddress = device.address;
+          Debug('_connect(): connectedAddress = ' + this.connectedAddress);
+          Debug('_connect(): onsuccess(): resolve');
+          resolve();
+        };
+
+        req.onerror = () => {
+          // No available profiles are connected.
+          Debug('_connect(): onerror(): reject with connection failed');
+          this.connectingAddress = null;
+          reject('connection failed');
+        };
+      });
+    },
+
+    /**
+     * The method will disconnect the connected device currently
+     * with the device's adapter.
+     *
+     * @access private
+     * @memberOf BluetoothConnectionManager
+     * @returns {Promise}
+     */
+    _disconnectCurrentConnectedDevice: 
+    function btcm__disconnectCurrentConnectedDevice() {
+      return new Promise((resolve, reject) => {
```

Comment:
It is not required to wrap the code in a promise. We could return the promise chain starting from `this._getConnectedDevices`.